### ISSUE TITLE
About TypeScript Support

### ISSUE THREAD

**Issue by toverux** [nodkz/graphql-compose#81]
Hi @nodkz!

At @mitmadness we're already using graphql-compose in a few projects and plan to use it in new ones that get a GraphQL API.

However, we're working with TypeScript, and there are no public definition files for graphql-compose and child libraries yet.
We've generated a low-quality stub with dts-gen, so at least we can `import` it, but not much more. No type checking/inference, no completion, no documentation, which makes the library quite hard to use sometimes.

So I'm considering creating a real, usable, complete declaration file for graphql-compose (and maybe other child libs), but since it's quite a lot of work and that it needs to be updated to follow the changes in your API, I wanted to know if you were interested in TS support or not.

If you are, we wouldn't have to publish the declaration on @types, since those files can be embedded in the library, which is the recommended solution. Also, when there are API changes / additions, the types could be updated right along with the JavaScript, to avoid static typing / runtime desyncs.
The downside is that it adds maintenance efforts, but I can help on that. We would need to add a few files more, though.

And last but not least: you seem to like Flow, but have you ever considered trying TS as well? Reading the source code, there wouln't be much changes in the library, and we would no longer need to maintain declaration files by hand (these are auto-generated by the TypeScript compiler), or need a "flow-typed" directory. You totally have the right to answer "TS is shit, Flow is much better at this and this or that", I won't be hurt =)

Anyway, thanks a lot for the work you've done around this set of libraries, these are quite awesome, and I've also seen how you're fast at answering issues and providing help. You're an excellent maintener and those graphql-compose-* libs deserve much more popularity, really! :+1:

**Comment by nodkz** [comment]
@toverux thanks a lot for your words and upcoming help.

> Yep, TS is "*shit" 😜. Flow is quite [powerful](https://twitter.com/thejameskyle/status/863195583191728129). But it burns a lot of my time for through checks between `graphql-js` and `graphql-compose`. So from other side Flow is "*shit" too. 😜
> PS. Please note that "*shit" means something with drawbacks or something that brings the pain in development.

I'm using Flowtype in `graphql-compose`, cause `graphql-js` uses it. It helps me to keep `graphql-compose` fully compatible with `graphql-js` and track any changes in its api and fix them in short time. I'm using React and Relay, they also covered by Flow, as result my internal app also under Flow. So my current stack with Facebook's packages enforces me to use Flow. And I'm quite happy now to use Flow. Pain was gone, now only happiness.

Flowtype and TypeScript are amazing tools for JS and ridiculously improve DX and code stability. It will be amazing if you add TS declaration files to `graphql-compose`. From my side, I try to keep them in sync with any upcoming changes in API.

PS. I do not like `flow-typed` (@types) approach for distributing type-defs. So feel free to open PRs  for including them directly to `graphql-compose`.

**Comment by toverux** [comment]
Thanks for your reactivity, as always! 

No problem, I've seen that you're more on the Babel/Flow side of the coin, those are cool tools too, and if you're happy with it, I'm happy too :smile:

I'll try to work on those type defs in the next few days (I'm not hurry, just wanted to test the waters), thanks a lot for accepting supporting TS (because oh yes - it make our lives easier!). Of course I'll still be around to help if needed.

Feel free to close the issue, the PR will do the rest.

**Comment by nodkz** [comment]
Let's keep this issue open for tracking progress.
When you will make commits, please fill them in the following format:
```
chore(typescript): Some commit message

Related #81
```

When all work will be done, we publish new `minor` version (2.**X**.0) via:
```
feat(typescript): Add typescript declarations

Closes #81
```
